### PR TITLE
Added SDL2 library

### PR DESCRIPTION
**Pull request by Cyberunner23**
Requested in #280.

### CODE REVIEW

**Comment by ruslo** [comment]
I've pushed some changes to `develop`, now it has conflicts and I was not able to merge it. Please rebase.

@Cyberunner23 Can you add an example so this package can be tested?

**Comment by Cyberunner23** [comment]
Rebase done, example should be up after some local tests.

**Comment by Cyberunner23** [comment]
Hmm... find_package is failing... What can be causing this?

```
CMake Error at CMakeLists.txt:10 (find_package):
  Could not find a package configuration file provided by "SDL2" with any of
  the following names:

    SDL2Config.cmake
    sdl2-config.cmake

  Add the installation prefix of "SDL2" to CMAKE_PREFIX_PATH or set
  "SDL2_DIR" to a directory containing one of the above files.  If "SDL2"
  provides a separate development package or SDK, be sure it has been
  installed.


-- Configuring incomplete, errors occurred!
See also "/home/cyberunner23/Dev/hunter/examples/SDL2/CMakeFiles/CMakeOutput.log".
```

**Comment by Cyberunner23** [comment]
I added the example in case you wanted to check it out.

**Comment by ruslo** [comment]
> What can be causing this?

Probably because it not supported by SDL2's CMake part. Can you show the link from docs about using installed SDL2 in user's CMake?

**Comment by Cyberunner23** [comment]
Seems like there is nothing in the SDL docs for using it with cmake though some people made a FindSDL2.cmake file, example usage [here](http://www.willusher.io/sdl2%20tutorials/2014/03/06/lesson-0-cmake/) and I also used SDL in another project directly by linking with the SDL targets [here](https://github.com/Cyberunner23/ProjectUnknown/blob/master/CMakeLists.txt)

**Comment by ruslo** [comment]
> Seems like there is nothing in the SDL docs for using it with cmake though some people made a FindSDL2.cmake file

It means we have two options now:
* Pick some `FindSDL2.cmake` and add it to the Hunter [cmake/find](https://github.com/ruslo/hunter/tree/master/cmake/find) modules (should be easy but not so good option)
* Patch CMake's SDL code so it will generate `SDL2Config.cmake` modules, see https://github.com/ruslo/hunter/wiki/example.find_package.config (require fork/patch, but it's the best)

**Comment by Cyberunner23** [comment]
Alright I'll have to work on this later but I would go with the patch.

**Comment by Cyberunner23** [comment]
We will have to patch it anyways because I just found I missed the part where it doesnt respect CMAKE_C_FLAGS...

**Comment by ruslo** [comment]
> We will have to patch it anyways

Does this project host sources on GitHub? Is there an official/unofficial GitHub fork already? There is a central place for patches: https://github.com/hunter-packages I can create repo and add you to contributors. Also it will be good to send this patches upstream so there will be no need to use "hunterized" fork.

**Comment by Cyberunner23** [comment]
There is an unofficial mirror on github located [here](https://github.com/spurious/SDL-mirror) it is an automatic mirror of the mercurial repo but there are no release tags on the github repo so it might be a bit hard to get the fork aligned with the releases.

**Comment by ruslo** [comment]
I've created repo https://github.com/hunter-packages/SDL2. Please keep `master` branch in a "clean" state so we can just fast-forward updates. Patches should be applied against branch `hunter` (which is default).

@Cyberunner23 check your email

> but there are no release tags on the github repo so it might be a bit hard to get the fork aligned with the releases

You can create them manually. Let me know if you have any better ideas.

**Comment by Cyberunner23** [comment]
Sorry for the delays, been deep in school work and I only managed to get time to work on this now. https://github.com/hunter-packages/SDL2 shoule be ready for testing.

**Comment by ruslo** [comment]
> shoule be ready for testing/inclusion

can you please update this pull request? thanks

**Comment by Cyberunner23** [comment]
Update what exactly?

**Comment by ruslo** [comment]
> Update what exactly?

**This** pull request. I'm not able to merge it because of conflicts (you can do rebase to fix it) also I see no new commits. Last commit in this pull request (Mar 3): https://github.com/ruslo/hunter/pull/336/commits/0378863ea396efc33091338e42b3d02dafb673dd

**Comment by Cyberunner23** [comment]
Rebase done, only thing left to do is to switch the project file to use the package from hunter-packages.

File: cmake/configs/default.cmake [hunk 1]
```diff
@@ -144,6 +144,8 @@ hunter_config(xproto VERSION 7.0.23)
 hunter_config(xrender VERSION 0.9.7)
 hunter_config(xshmfence VERSION 1.2)
 hunter_config(xtrans VERSION 1.2.7)
+hunter_config(ZeroMQ VERSION 4.1.4)
```
Comment (ruslo) [inline cmake/configs/default.cmake]: I think this one is not relate to `SDL2`

File: cmake/projects/SDL2/hunter.cmake [hunk 2]
```diff
@@ -0,0 +1,24 @@
+if(DEFINED HUNTER_CMAKE_PROJECTS_SDL2_HUNTER_CMAKE_)
```
Comment (ruslo) [inline cmake/projects/SDL2/hunter.cmake]: Header guards may lead to wrong configuration in some cases see wiki:
* https://github.com/ruslo/hunter/wiki/usr.adding.new.package#03-add-new-package

please add comment `# !!! DO NOT PLACE HEADER GUARDS HERE !!!`

File: cmake/projects/SDL2/hunter.cmake [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    SDL2
+    VERSION
+    "2.0.4-h1"
+    URL
+    "https://github.com/hunter-packages/SDL2/archive/2.0.4-h1.tar.gz"
+    SHA1
+    632a4005aafb25d6fc894a93475f78832a1edd98
+)
+
+hunter_pick_scheme(DEFAULT url_sha1_cmake)
```
Comment (ruslo) [inline cmake/projects/SDL2/hunter.cmake]: Is this package cacheable? If so please add `hunter_cacheable` (see [wiki](https://github.com/ruslo/hunter/wiki/usr.adding.new.package#important))

File: cmake/configs/default.cmake (hunk 1 above)
Comment (Cyberunner23) [inline cmake/configs/default.cmake]: Done, must of missed it when it popped up in the rebase.

File: cmake/projects/SDL2/hunter.cmake (hunk 2 above)
Comment (Cyberunner23) [inline cmake/projects/SDL2/hunter.cmake]: Done

File: cmake/projects/SDL2/hunter.cmake (hunk 3 above)
Comment (Cyberunner23) [inline cmake/projects/SDL2/hunter.cmake]: Done

File: cmake/projects/SDL2/hunter.cmake [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# !!! DO NOT PLACE HEADER GUARDS HERE !!!
+
+# Load used modules
+include(hunter_add_version)
+include(hunter_download)
+include(hunter_pick_scheme)
+
```
Comment (ruslo) [inline cmake/projects/SDL2/hunter.cmake]: `include(hunter_cacheable)` needed

Comment (Cyberunner23) [inline cmake/projects/SDL2/hunter.cmake]: Done

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.479
* https://travis-ci.org/ingenue/hunter/builds/135259615

**Comment by ruslo** [comment]
> Testing

```
/home/travis/build/ingenue/hunter/examples/SDL2/main.cpp:2:9: fatal error: 'SDL.h' file not found
#include<SDL.h>
        ^
1 error generated.
```

I think it shoud be `#include <SDL2/SDL.h>` (?)

**Comment by Cyberunner23** [comment]
Ready for retest.

**Comment by Cyberunner23** [comment]
Tests work locally, should be fine to do the tests on travos and appveyor

File: examples/SDL2/main.cpp [hunk 5]
```diff
@@ -0,0 +1,15 @@
+#include<iostream>
+#include<SDL/SDL.h>
```
Comment (ruslo) [inline examples/SDL2/main.cpp]: > Tests work locally, should be fine to do the tests on travis and appveyor/merge

not for me, I think it should be `#include <SDL2/SDL.h>`

Comment (Cyberunner23) [inline examples/SDL2/main.cpp]: Tried with `#include <SDL2/SDL.h>` and broke... Quite strange... will take a look.

Comment (Cyberunner23) [inline examples/SDL2/main.cpp]: I don't understand, `#include <SDL2/SDL.h>` breaks and `#include <SDL/SDL.h` works fine for me, can I take a look at your test perhaps?

Comment (Cyberunner23) [inline examples/SDL2/main.cpp]: I dont get it... `#include <SDL2/SDL.h>` still breaks for me...

Comment (Cyberunner23) [inline examples/SDL2/main.cpp]: Which test are you using?

Comment (ruslo) [inline examples/SDL2/main.cpp]: > Which test are you using?

https://github.com/ruslo/hunter/wiki/dev.hunter.root#jenkinspy

Comment (Cyberunner23) [inline examples/SDL2/main.cpp]: Ran `TOOLCHAIN=default PROJECT_DIR=examples/SDL2 ./jenkins.py --verbose` with `#include <SDL/SDL.h>` and it worked fine... I don't know what to say...

Comment (ruslo) [inline examples/SDL2/main.cpp]: > I don't know what to say

probably this package installed system-wide? what OS are you using?

Comment (Cyberunner23) [inline examples/SDL2/main.cpp]: Arch linux.

Comment (Cyberunner23) [inline examples/SDL2/main.cpp]: Give me a sec I'll test it on something that doesnt have it, not even SDL1.2 (which is the only system wide SDL that I have)

Comment (ruslo) [inline examples/SDL2/main.cpp]: also you can use the power of docker: https://github.com/hunter-packages/travis-linux-docker

Comment (Cyberunner23) [inline examples/SDL2/main.cpp]: Ended up trying it on travis, getting internal errors though https://travis-ci.org/Cyberunner23/CITests/builds/135431759

Comment (ruslo) [inline examples/SDL2/main.cpp]: > getting internal errors though

it's a Hunter issue, will fix it now. you can simply re-run it again

Comment (ruslo) [inline examples/SDL2/main.cpp]: > will fix it now

pushed to `develop`: https://github.com/ruslo/hunter/commit/964a87e3e0f1af6dade2c13a1007274c39fa37a2

Comment (ruslo) [inline examples/SDL2/main.cpp]: > you can simply re-run it again

I see it doesn't help for you. The error is about CMake was not able to handle `https`. You can use instructions similar to Hunter testing to download appropriate version:
* https://github.com/ruslo/hunter/blob/0380dbd81b80007f4c9c3d01e309d9820a833619/.travis.yml#L74

Comment (Cyberunner23) [inline examples/SDL2/main.cpp]: ok umm install-ci-dependencies.py isnt working either

Comment (Cyberunner23) [inline examples/SDL2/main.cpp]: Still happening with the latest version of CMake...

Comment (Cyberunner23) [inline examples/SDL2/main.cpp]: If it helps I checked what its trying to download (https://raw.githubusercontent.com/ingenue/hunter-cache/master/e77cb10/SDL2/2.0.4-h1/632a400/da39a3e/a49b0e5/da39a3e/basic-deps.DONE) and its just says "Not Found"

Comment (ruslo) [inline examples/SDL2/main.cpp]: > Still happening with the latest version of CMake...

can you show logs?

Comment (Cyberunner23) [inline examples/SDL2/main.cpp]: https://travis-ci.org/Cyberunner23/CITests/builds/135443811

Comment (ruslo) [inline examples/SDL2/main.cpp]: > If it helps I checked what its trying to download...

it's okay, it means "cache miss" (error `22`) => Hunter need to buid package from sources. Error `1` mean CMake internal error.

Comment (Cyberunner23) [inline examples/SDL2/main.cpp]: ohh ok, still can't figure out why it keeps erroring with "Unsupportes protocol"...

**Comment by Cyberunner23** [comment]
also fails with `#include<SDL2/SDL.h>`...

**Comment by Cyberunner23** [comment]
Include propagation fixed, Local tests work, test in travis (https://travis-ci.org/Cyberunner23/CITests/builds/135450613) work

**Comment by ruslo** [comment]
testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.480
* https://travis-ci.org/ingenue/hunter/builds/135462205

**Comment by ruslo** [comment]
> testing

some tests failed, need to fix/exclude broken stuff

**Comment by Cyberunner23** [comment]
Closing PR, moved SDL2 package to develop_sdl2 branch, havinf SDL2 in develop was causing issues when branching off to work on other packages. New PR: #441

**Pull request closed without merging**